File: string-art-python/string_art.py
#!/usr/bin/env python3
"""
String Art Generator - Python implementation optimized for speed on macOS
Ported from Go version with NumPy and Numba optimizations
"""

import argparse
import time
import math
import numpy as np
from PIL import Image
import aggdraw
from numba import jit
from typing import List, Tuple, Optional


# Constants
MIN_DISTANCE = 30


@jit(nopython=True)
def calculate_pin_coords(pins: int, img_size: int) -> np.ndarray:
    """Calculate pin coordinates around a circle."""
    pin_coords = np.zeros((pins, 2), dtype=np.float64)
    center = img_size / 2
    radius = img_size / 2 - 1

    for i in range(pins):
        angle = 2 * math.pi * i / pins
        pin_coords[i, 0] = math.floor(center + radius * math.cos(angle))  # X
        pin_coords[i, 1] = math.floor(center + radius * math.sin(angle))  # Y

    return pin_coords


@jit(nopython=True)
def bresenham_line(x0: int, y0: int, x1: int, y1: int) -> Tuple[np.ndarray, np.ndarray]:
    """Generate line coordinates using Bresenham's algorithm."""
    dx = abs(x1 - x0)
    dy = abs(y1 - y0)
    sx = 1 if x0 < x1 else -1
    sy = 1 if y0 < y1 else -1
    err = dx - dy

    # Estimate maximum number of points
    max_points = dx + dy + 1
    x_coords = np.zeros(max_points, dtype=np.int32)
    y_coords = np.zeros(max_points, dtype=np.int32)

    x, y = x0, y0
    idx = 0

    while True:
        x_coords[idx] = x
        y_coords[idx] = y
        idx += 1

        if x == x1 and y == y1:
            break

        e2 = 2 * err
        if e2 > -dy:
            err -= dy
            x += sx
        if e2 < dx:
            err += dx
            y += sy

    return x_coords[:idx], y_coords[:idx]


def precalculate_lines(pin_coords: np.ndarray, pins: int):
    """Precalculate all potential lines between pins."""
    # Store line data in a dictionary for efficient lookup
    line_cache = {}

    print("Precalculating potential lines...")
    for i in range(pins):
        if i % 50 == 0:
            print(f"  Processing pin {i}/{pins}")

        for j in range(i + MIN_DISTANCE, pins):
            x0 = int(pin_coords[i, 0])
            y0 = int(pin_coords[i, 1])
            x1 = int(pin_coords[j, 0])
            y1 = int(pin_coords[j, 1])

            xs, ys = bresenham_line(x0, y0, x1, y1)

            # Store both directions
            line_cache[(i, j)] = (xs, ys)
            line_cache[(j, i)] = (xs, ys)

    return line_cache


@jit(nopython=True)
def get_line_error(
    error_img: np.ndarray, x_coords: np.ndarray, y_coords: np.ndarray, img_size: int
) -> float:
    """Calculate the error reduction for a line."""
    total_error = 0.0
    valid_points = 0

    for i in range(len(x_coords)):
        x, y = x_coords[i], y_coords[i]
        if 0 <= x < img_size and 0 <= y < img_size:
            total_error += error_img[y, x]
            valid_points += 1

    return total_error / valid_points if valid_points > 0 else 0.0


@jit(nopython=True)
def apply_line_to_error(
    error_img: np.ndarray,
    x_coords: np.ndarray,
    y_coords: np.ndarray,
    line_weight: int,
    img_size: int,
):
    """Apply a line to the error image."""
    for i in range(len(x_coords)):
        x, y = x_coords[i], y_coords[i]
        if 0 <= x < img_size and 0 <= y < img_size:
            error_img[y, x] -= line_weight


@jit(nopython=True)
def contains_recent_pin(last_pins: np.ndarray, pin: int) -> bool:
    """Check if a pin was used recently."""
    for i in range(len(last_pins)):
        if last_pins[i] == pin:
            return True
    return False


class StringArtGenerator:
    def __init__(
        self,
        pins: int = 300,
        max_lines: int = 4000,
        target_size: int = 500,
        output_size: Optional[int] = None,
        line_weight: int = 8,
        output_weight: Optional[int] = None,
    ):
        self.pins = pins
        self.max_lines = max_lines
        self.target_size = target_size
        self.output_size = output_size or target_size
        self.line_weight = line_weight
        self.output_weight = output_weight or line_weight

        self.pin_coords = None
        self.line_cache = None

    def load_and_process_image(self, filename: str) -> np.ndarray:
        """Load and process input image."""
        print(f"Processing {filename}...")

        # Load image
        img = Image.open(filename).convert("RGB")
        orig_width, orig_height = img.size

        # Crop to square and resize
        size = min(orig_width, orig_height)
        left = (orig_width - size) // 2
        top = (orig_height - size) // 2
        img = img.crop((left, top, left + size, top + size))
        img = img.resize((self.target_size, self.target_size), Image.Resampling.LANCZOS)

        print(
            f"  Resized from {orig_width}x{orig_height} to {self.target_size}x{self.target_size}"
        )

        # Convert to grayscale using luminance formula
        img_array = np.array(img, dtype=np.float64)
        luminance = (
            0.2126 * img_array[:, :, 0]
            + 0.7152 * img_array[:, :, 1]
            + 0.0722 * img_array[:, :, 2]
        )

        return luminance

    def calculate_lines(self, source_image: np.ndarray) -> List[int]:
        """Calculate the optimal sequence of lines."""
        print("Calculating string art lines...")

        # Initialize error image (inverted)
        error = 255.0 - source_image

        # Initialize line sequence
        line_sequence = [0]  # Start from pin 0
        current_pin = 0
        last_pins = np.full(20, -1, dtype=np.int32)  # Track recent pins
        last_pin_idx = 0

        for i in range(self.max_lines):
            if i % 100 == 0:
                print(f"  Processing line {i}/{self.max_lines}")

            best_pin = -1
            max_error = 0.0
            best_x_coords = None
            best_y_coords = None

            # Try all possible pins
            for offset in range(MIN_DISTANCE, self.pins - MIN_DISTANCE):
                test_pin = (current_pin + offset) % self.pins

                # Skip if recently used
                if contains_recent_pin(last_pins, test_pin):
                    continue

                # Get line coordinates
                x_coords, y_coords = self.line_cache[(current_pin, test_pin)]

                # Calculate error reduction
                line_error = get_line_error(error, x_coords, y_coords, self.target_size)

                if line_error > max_error:
                    max_error = line_error
                    best_pin = test_pin
                    best_x_coords = x_coords
                    best_y_coords = y_coords

            if best_pin == -1:
                # No valid pin found
                break

            # Add best pin to sequence
            line_sequence.append(best_pin)

            # Apply line to error image
            apply_line_to_error(
                error, best_x_coords, best_y_coords, self.line_weight, self.target_size
            )

            # Update recent pins tracking
            last_pins[last_pin_idx] = best_pin
            last_pin_idx = (last_pin_idx + 1) % len(last_pins)
            current_pin = best_pin

        print()
        return line_sequence

    def generate_output_image(self, line_sequence: List[int], output_file: str):
        """Generate the final string art image with anti-aliased lines using aggdraw."""
        print("Generating output image with anti-aliased lines...")

        # Create white image
        img = Image.new("RGB", (self.output_size, self.output_size), "white")

        # Create aggdraw canvas for anti-aliased drawing
        canvas = aggdraw.Draw(img)

        # Calculate scale factor
        scale = self.output_size / self.target_size

        # Create pens and brushes for drawing
        circle_pen = aggdraw.Pen("black", 1)
        pin_brush = aggdraw.Brush("black")

        # Calculate line opacity (0-255 range for aggdraw)
        line_opacity = min(255, max(1, self.output_weight))
        line_pen = aggdraw.Pen("rgba(0,0,0)", 1, line_opacity)

        # Draw circle border
        center = self.output_size // 2
        radius = self.output_size // 2 - 1
        canvas.ellipse(
            (center - radius, center - radius, center + radius, center + radius),
            circle_pen,
        )

        # Draw pins
        for coord in self.pin_coords:
            x = int(coord[0] * scale)
            y = int(coord[1] * scale)
            canvas.ellipse((x - 2, y - 2, x + 2, y + 2), circle_pen, pin_brush)

        # Draw anti-aliased lines
        print(f"  Drawing {len(line_sequence) - 1} anti-aliased lines...")
        for i in range(len(line_sequence) - 1):
            if i % 1000 == 0 and i > 0:
                print(f"    Drawing line {i}/{len(line_sequence) - 1}")

            from_pin = line_sequence[i]
            to_pin = line_sequence[i + 1]

            from_coord = self.pin_coords[from_pin]
            to_coord = self.pin_coords[to_pin]

            x0 = int(from_coord[0] * scale)
            y0 = int(from_coord[1] * scale)
            x1 = int(to_coord[0] * scale)
            y1 = int(to_coord[1] * scale)

            canvas.line((x0, y0, x1, y1), line_pen)

        # Flush the canvas to apply all drawing operations
        canvas.flush()

        # Save image
        img.save(output_file, "PNG")
        print(f"Output saved to {output_file}")

    def generate(self, input_file: str, output_file: str = "output.png"):
        """Generate string art from input image."""
        print(
            f"  Pins: {self.pins}, Max lines: {self.max_lines}, Processing size: {self.target_size}"
        )
        print(
            f"  Output size: {self.output_size}, Line weight: {self.line_weight}, Output weight: {self.output_weight}"
        )

        # Load and process image
        source_image = self.load_and_process_image(input_file)

        start_time = time.time()

        # Calculate pin coordinates
        self.pin_coords = calculate_pin_coords(self.pins, self.target_size)

        # Precalculate lines
        self.line_cache = precalculate_lines(self.pin_coords, self.pins)

        # Calculate optimal line sequence
        line_sequence = self.calculate_lines(source_image)

        end_time = time.time()
        print(f"Processing took {end_time - start_time:.2f} seconds")

        # Generate output image
        self.generate_output_image(line_sequence, output_file)


def main():
    parser = argparse.ArgumentParser(description="Generate string art from an image")
    parser.add_argument("--input", required=True, help="Input image file (JPEG or PNG)")
    parser.add_argument("--output", default="output.png", help="Output image file")
    parser.add_argument(
        "--pins", type=int, default=300, help="Number of pins around the circle"
    )
    parser.add_argument(
        "--lines", type=int, default=4000, help="Maximum number of lines to draw"
    )
    parser.add_argument(
        "--size",
        type=int,
        default=500,
        help="Processing size for input image in pixels",
    )
    parser.add_argument(
        "--output-size",
        type=int,
        help="Output image size in pixels (default: same as processing size)",
    )
    parser.add_argument(
        "--weight", type=int, default=8, help="Line weight for darkness calculation"
    )
    parser.add_argument(
        "--output-weight",
        type=int,
        help="Visual line opacity for output image (default: same as weight)",
    )

    args = parser.parse_args()

    generator = StringArtGenerator(
        pins=args.pins,
        max_lines=args.lines,
        target_size=args.size,
        output_size=args.output_size,
        line_weight=args.weight,
        output_weight=args.output_weight,
    )

    generator.generate(args.input, args.output)


if __name__ == "__main__":
    main()

File: stringart.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
#include <chrono>
#include <cstring>
#include <unordered_set>

#ifdef _OPENMP
#include <omp.h>
#endif

#define STB_IMAGE_IMPLEMENTATION
#include "stb_image.h"
#define STB_IMAGE_WRITE_IMPLEMENTATION
#include "stb_image_write.h"
#define STB_IMAGE_RESIZE_IMPLEMENTATION
#include "stb_image_resize.h"

struct Coord {
    double x;
    double y;
};

struct Config {
    int pins = 300;
    int maxLines = 4000;
    int targetSize = 500;
    int outputSize = 500;
    int lineWeight = 8;
    int outputWeight = 8;
    int minDistance = 30;
    std::string inputFile;
    std::string outputFile = "output.png";
};

class StringArtGenerator {
private:
    Config config;
    int imgSize;
    double imgSizeF;
    int imgSizeSq;

    std::vector<Coord> pinCoords;
    std::vector<double> sourceImage;
    std::vector<std::vector<int>> lineCacheX;
    std::vector<std::vector<int>> lineCacheY;

public:
    StringArtGenerator(const Config& cfg) : config(cfg) {
        imgSize = config.targetSize;
        imgSizeF = static_cast<double>(imgSize);
        imgSizeSq = imgSize * imgSize;
    }

    void run() {
        std::cout << "Processing " << config.inputFile << "...\n";
        std::cout << "  Pins: " << config.pins << ", Max lines: " << config.maxLines
                  << ", Processing size: " << config.targetSize
                  << ", Output size: " << config.outputSize
                  << ", Line weight: " << config.lineWeight
                  << ", Output weight: " << config.outputWeight
                  << ", Min distance: " << config.minDistance << "\n";

        auto start = std::chrono::high_resolution_clock::now();

        loadAndProcessImage();
        calculatePinCoords();
        precalculateAllPotentialLines();
        auto lineSequence = calculateLines();

        auto end = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double> diff = end - start;
        std::cout << "Processing took " << diff.count() << " seconds\n";

        std::cout << "Generating output image...\n";
        generateOutputImage(lineSequence);
        std::cout << "Output saved to " << config.outputFile << "\n";
    }

private:
    void loadAndProcessImage() {
        int width, height, channels;
        unsigned char* img = stbi_load(config.inputFile.c_str(), &width, &height, &channels, 0);

        if (!img) {
            std::cerr << "Error loading image: " << config.inputFile << "\n";
            exit(1);
        }

        // Find minimum dimension and crop to square
        int size = std::min(width, height);
        int startX = (width - size) / 2;
        int startY = (height - size) / 2;

        // Create temporary cropped image
        std::vector<unsigned char> cropped(size * size * channels);
        for (int y = 0; y < size; ++y) {
            std::memcpy(&cropped[y * size * channels],
                       &img[(startY + y) * width * channels + startX * channels],
                       size * channels);
        }

        // Resize to target size
        std::vector<unsigned char> resized(imgSize * imgSize * channels);
        stbir_resize_uint8(&cropped[0], size, size, 0,
                          &resized[0], imgSize, imgSize, 0, channels);

        if (width != config.targetSize || height != config.targetSize) {
            std::cout << "  Resized from " << width << "x" << height
                     << " to " << imgSize << "x" << imgSize << "\n";
        }

        // Convert to grayscale luminosity
        sourceImage.resize(imgSizeSq);

        #ifdef _OPENMP
        #pragma omp parallel for
        #endif
        for (int i = 0; i < imgSizeSq; ++i) {
            int idx = i * channels;
            double r = resized[idx];
            double g = (channels > 1) ? resized[idx + 1] : r;
            double b = (channels > 2) ? resized[idx + 2] : r;

            // ITU-R BT.709 luminosity formula
            sourceImage[i] = 0.2126 * r + 0.7152 * g + 0.0722 * b;
        }

        stbi_image_free(img);
    }

    void calculatePinCoords() {
        pinCoords.resize(config.pins);
        double center = imgSize / 2.0;
        double radius = imgSize / 2.0 - 1;

        for (int i = 0; i < config.pins; ++i) {
            double angle = 2 * M_PI * i / config.pins;
            pinCoords[i].x = std::floor(center + radius * std::cos(angle));
            pinCoords[i].y = std::floor(center + radius * std::sin(angle));
        }
    }

    void precalculateAllPotentialLines() {
        int cacheSize = config.pins * config.pins;
        lineCacheX.resize(cacheSize);
        lineCacheY.resize(cacheSize);

        #ifdef _OPENMP
        #pragma omp parallel for collapse(2)
        #endif
        for (int i = 0; i < config.pins; ++i) {
            for (int j = i + config.minDistance; j < config.pins; ++j) {
                double x0 = pinCoords[i].x;
                double y0 = pinCoords[i].y;
                double x1 = pinCoords[j].x;
                double y1 = pinCoords[j].y;

                double dx = x1 - x0;
                double dy = y1 - y0;
                int steps = static_cast<int>(std::sqrt(dx * dx + dy * dy));

                std::vector<int> xs(steps);
                std::vector<int> ys(steps);

                for (int k = 0; k < steps; ++k) {
                    double t = static_cast<double>(k) / (steps - 1);
                    xs[k] = static_cast<int>(x0 + t * dx);
                    ys[k] = static_cast<int>(y0 + t * dy);
                }

                int idx1 = j * config.pins + i;
                int idx2 = i * config.pins + j;

                lineCacheX[idx1] = xs;
                lineCacheX[idx2] = xs;
                lineCacheY[idx1] = ys;
                lineCacheY[idx2] = ys;
            }
        }
    }

    double getLineError(const std::vector<double>& error, int cacheIndex) {
        const auto& xs = lineCacheX[cacheIndex];
        const auto& ys = lineCacheY[cacheIndex];

        if (xs.empty()) return 0.0;

        double sum = 0.0;
        for (size_t i = 0; i < xs.size(); ++i) {
            sum += error[ys[i] * imgSize + xs[i]];
        }
        return sum / xs.size();
    }

    std::vector<int> calculateLines() {
        std::cout << "Calculating string art lines...\n";

        std::vector<double> error(imgSizeSq);
        #ifdef _OPENMP
        #pragma omp parallel for
        #endif
        for (int i = 0; i < imgSizeSq; ++i) {
            error[i] = 255.0 - sourceImage[i];
        }

        std::vector<int> lineSequence;
        lineSequence.reserve(config.maxLines + 1);
        lineSequence.push_back(0);

        int currentPin = 0;
        std::vector<int> lastPins(20, -1);

        for (int i = 0; i < config.maxLines; ++i) {
            if (i % 100 == 0) {
                std::cout << "  Processing line " << i << "/" << config.maxLines << "\r" << std::flush;
            }

            int bestPin = -1;
            double maxErr = 0.0;
            int bestIndex = 0;

            // Find best next pin
            for (int offset = config.minDistance; offset < config.pins - config.minDistance; ++offset) {
                int testPin = (currentPin + offset) % config.pins;

                // Check if pin was recently used
                bool skip = false;
                for (int recentPin : lastPins) {
                    if (recentPin == testPin) {
                        skip = true;
                        break;
                    }
                }
                if (skip) continue;

                int index = testPin * config.pins + currentPin;
                double lineErr = getLineError(error, index);

                if (lineErr > maxErr) {
                    maxErr = lineErr;
                    bestPin = testPin;
                    bestIndex = index;
                }
            }

            if (bestPin == -1) break;

            lineSequence.push_back(bestPin);

            // Update error map
            const auto& xs = lineCacheX[bestIndex];
            const auto& ys = lineCacheY[bestIndex];
            for (size_t j = 0; j < xs.size(); ++j) {
                int idx = ys[j] * imgSize + xs[j];
                error[idx] -= config.lineWeight;
            }

            // Update recent pins list
            for (int j = 0; j < 19; ++j) {
                lastPins[j] = lastPins[j + 1];
            }
            lastPins[19] = bestPin;
            currentPin = bestPin;
        }

        std::cout << "\n";
        return lineSequence;
    }

    void generateOutputImage(const std::vector<int>& lineSequence) {
        int outSize = config.outputSize;
        std::vector<unsigned char> image(outSize * outSize * 4, 255);

        // Set alpha channel
        for (int i = 3; i < outSize * outSize * 4; i += 4) {
            image[i] = 255;
        }

        double scale = static_cast<double>(outSize) / imgSize;

        // Draw circle border
        int centerOut = outSize / 2;
        int radiusOut = outSize / 2 - 1;
        drawCircle(image, centerOut, centerOut, radiusOut, outSize);

        // Draw pins
        for (const auto& coord : pinCoords) {
            int scaledX = static_cast<int>(coord.x * scale);
            int scaledY = static_cast<int>(coord.y * scale);
            drawFilledCircle(image, scaledX, scaledY, 2, outSize);
        }

        // Draw lines with alpha blending
        for (size_t i = 0; i < lineSequence.size() - 1; ++i) {
            const auto& from = pinCoords[lineSequence[i]];
            const auto& to = pinCoords[lineSequence[i + 1]];

            int x0 = static_cast<int>(from.x * scale);
            int y0 = static_cast<int>(from.y * scale);
            int x1 = static_cast<int>(to.x * scale);
            int y1 = static_cast<int>(to.y * scale);

            drawLineAlpha(image, x0, y0, x1, y1, config.outputWeight, outSize);
        }

        stbi_write_png(config.outputFile.c_str(), outSize, outSize, 4, image.data(), outSize * 4);
    }

    void drawLineAlpha(std::vector<unsigned char>& img, int x0, int y0, int x1, int y1, int alpha, int size) {
        int dx = std::abs(x1 - x0);
        int dy = std::abs(y1 - y0);
        int sx = (x0 < x1) ? 1 : -1;
        int sy = (y0 < y1) ? 1 : -1;
        int err = dx - dy;

        double alphaF = alpha / 255.0;
        double invAlpha = 1.0 - alphaF;

        while (true) {
            if (x0 >= 0 && x0 < size && y0 >= 0 && y0 < size) {
                int idx = (y0 * size + x0) * 4;
                img[idx] = static_cast<unsigned char>(img[idx] * invAlpha);
                img[idx + 1] = static_cast<unsigned char>(img[idx + 1] * invAlpha);
                img[idx + 2] = static_cast<unsigned char>(img[idx + 2] * invAlpha);
            }

            if (x0 == x1 && y0 == y1) break;

            int e2 = 2 * err;
            if (e2 > -dy) {
                err -= dy;
                x0 += sx;
            }
            if (e2 < dx) {
                err += dx;
                y0 += sy;
            }
        }
    }

    void drawCircle(std::vector<unsigned char>& img, int cx, int cy, int r, int size) {
        for (double angle = 0; angle < 2 * M_PI; angle += 0.01) {
            int x = cx + static_cast<int>(r * std::cos(angle));
            int y = cy + static_cast<int>(r * std::sin(angle));
            if (x >= 0 && x < size && y >= 0 && y < size) {
                int idx = (y * size + x) * 4;
                img[idx] = img[idx + 1] = img[idx + 2] = 0;
            }
        }
    }

    void drawFilledCircle(std::vector<unsigned char>& img, int cx, int cy, int r, int size) {
        for (int x = -r; x <= r; ++x) {
            for (int y = -r; y <= r; ++y) {
                if (x * x + y * y <= r * r) {
                    int px = cx + x;
                    int py = cy + y;
                    if (px >= 0 && px < size && py >= 0 && py < size) {
                        int idx = (py * size + px) * 4;
                        img[idx] = img[idx + 1] = img[idx + 2] = 0;
                    }
                }
            }
        }
    }
};

int main(int argc, char* argv[]) {
    Config config;

    // Parse command line arguments
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];

        if (arg == "-input" && i + 1 < argc) {
            config.inputFile = argv[++i];
        } else if (arg == "-output" && i + 1 < argc) {
            config.outputFile = argv[++i];
        } else if (arg == "-pins" && i + 1 < argc) {
            config.pins = std::stoi(argv[++i]);
        } else if (arg == "-lines" && i + 1 < argc) {
            config.maxLines = std::stoi(argv[++i]);
        } else if (arg == "-size" && i + 1 < argc) {
            config.targetSize = std::stoi(argv[++i]);
        } else if (arg == "-output-size" && i + 1 < argc) {
            config.outputSize = std::stoi(argv[++i]);
        } else if (arg == "-weight" && i + 1 < argc) {
            config.lineWeight = std::stoi(argv[++i]);
        } else if (arg == "-output-weight" && i + 1 < argc) {
            config.outputWeight = std::stoi(argv[++i]);
        } else if (arg == "-min-distance" && i + 1 < argc) {
            config.minDistance = std::stoi(argv[++i]);
        }
    }

    if (config.outputSize == 0) {
        config.outputSize = config.targetSize;
    }
    if (config.outputWeight == 0) {
        config.outputWeight = config.lineWeight;
    }

    if (config.inputFile.empty()) {
        std::cerr << "Please provide an input file using -input flag\n";
        return 1;
    }

    StringArtGenerator generator(config);
    generator.run();

    return 0;
}